compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -1439,6 +1439,14 @@ public static BType getReferredType(BType type) {
         return constraint;
     }
 
+    public static BType getEffectiveType(BType type) {
+        type = Types.getReferredType(type);
+        if (type.tag == TypeTags.INTERSECTION) {
+            return ((BIntersectionType) type).effectiveType;
+        }
+        return type;
```

Comment:
Shouldn't this return the input type if it's not an intersection? If you meant to unwrap both reference and intersection type; rename the function name.

Reply:
Updated the `getEffectiveType` function